app/controllers/case_contacts/form_controller.rb

```diff
@@ -0,0 +1,145 @@
+      end
+    end
+  end
+
+  private
+
+  def get_cases_and_contact_types
+    @casa_cases = policy_scope(current_organization.casa_cases)
+    @casa_cases = @casa_cases.where(id: @case_contact.casa_case_id) if @case_contact.active?
+
+    @selected_case_contact_types = @casa_cases.flat_map(&:contact_types)
+
+    @current_organization_groups =
+      if @selected_case_contact_types.present?
+        @selected_case_contact_types.map(&:contact_type_group).uniq
+      else
+        current_organization.contact_types_by_group
+      end
+  end
+
+  def finish_editing
+    message = ""
+    send_reimbursement_email(@case_contact)
+    if @case_contact.active?
+      message = @case_contact.decorate.form_updated_message
+    else
+      message = "Case #{"contact".pluralize(@case_contact.draft_case_ids.count)} successfully created."
+      create_additional_case_contacts(@case_contact)
+      first_casa_case_id = @case_contact.draft_case_ids.slice(0)
+      @case_contact.update(status: "active", draft_case_ids: [first_casa_case_id], casa_case_id: first_casa_case_id)
+    end
+    update_volunteer_address(@case_contact)
+    flash[:notice] = message
+    redirect_back_to_referer(fallback_location: case_contacts_path(success: true))
+  end
+
+  def send_reimbursement_email(case_contact)
+    if case_contact.should_send_reimbursement_email?
+      SupervisorMailer.reimbursement_request_email(case_contact.creator, case_contact.supervisor).deliver_later
+    end
+  end
+
+  def update_volunteer_address(case_contact)
+    return unless case_contact.volunteer_address.present? && !case_contact.address_field_disabled?
+
+    address = case_contact.volunteer.address || case_contact.volunteer.build_address
+    address.update(content: case_contact.volunteer_address)
+  end
+
+  # Makes a copy of the draft for all selected cases not including the first one. The draft becomes the contact for
+  # the first case.
+  #
+  # Duplication does not duplicate associated records, so if other associations are made in the form, they need to be
+  # added here, explicitly (ie. case_contact_contact_type, additional_expenses). Alternatively, could look at a gem
+  # that does deep associations.
+  def create_additional_case_contacts(case_contact)
+    case_contact.draft_case_ids.drop(1).each do |casa_case_id|
+      new_case_contact = case_contact.dup
+      new_case_contact.status = "active"
+      new_case_contact.draft_case_ids = [casa_case_id]
+      new_case_contact.casa_case_id = casa_case_id
+      case_contact.case_contact_contact_type.each do |ccct|
+        new_case_contact.case_contact_contact_type.new(contact_type_id: ccct.contact_type_id)
+      end
+      case_contact.additional_expenses.each do |ae|
+        new_case_contact.additional_expenses.new(
+          other_expense_amount: ae.other_expense_amount,
+          other_expenses_describe: ae.other_expenses_describe
+        )
+      end
+      new_case_contact.save!
```

Comment:
Can we make this:

```
unless new_case_contact.save
  new_case_contact.errors.add(...some error)
end
```

or will that not work in the UI?

Reply:
This is also past the point of no return. This is the duplication process, so it's also a clone of a case contact that has passed validation. Happy to take the bang off, but we wouldn't want validation errors here.

This test makes sure the cloning works. See the block on 289.
spec/requests/case_contacts/form_spec.rb